src/Algebra/Graph.hs

```diff
@@ -1187,68 +1288,69 @@ sparsifyKL n graph = KL.buildG (1, next - 1) ((n + 1, n + 2) : Exts.toList (res
 The rules for foldg work very similarly to GHC's mapFB rules; see a note below
 this line: http://hackage.haskell.org/package/base/docs/src/GHC.Base.html#mapFB.
 
-* Up to (but not including) phase 1, we use the "buildR/f" rule to rewrite all
-  saturated applications of f into its buildR/foldg form, hoping for fusion to
-  happen (through the "foldg/buildR" rule).
+* Up to (but not including) phase 1, we use the "buildg/f" rule to rewrite all
+  saturated applications of f into its buildg/foldg form, hoping for fusion to
+  happen (through the "foldg/buildg" rule).
 
-  In phases 1 and 0, we switch off these rules, inline buildR, and switch on the
+  In phases 1 and 0, we switch off these rules, inline buildg, and switch on the
   "graph/f" rule, which rewrites "foldg/f" back into plain functions if needed.
 
   It's important that these two rules aren't both active at once (along with
   build's unfolding) else we'd get an infinite loop in the rules. Hence the
   activation control below.
 
 * composeR and matchR are here to remember the original function after applying
-  a "buildR/f" rule. These functions are higher-order functions and therefore
+  a "buildg/f" rule. These functions are higher-order functions and therefore
   benefit from inlining in the final phase.
 
 * The "bindR/bindR" rule optimises compositions of multiple bindR's.
 -}
 
-type Foldg a = forall b. b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> b
-
-buildR :: Foldg a -> Graph a
-buildR g = g Empty Vertex Overlay Connect
-{-# INLINE [1] buildR #-}
+-- | Build a graph given an interpretation of the four graph construction primitives 'empty',
+-- 'vertex', 'overlay' and 'connect', in this order. See examples for further clarification.
+--
+-- Functions expressed with 'buildg' are good producers.
+--
+-- @
+-- buildg f                                                   == f 'empty' 'vertex' 'overlay' 'connect'
+-- buildg (\\e _ _ _ -> e)                                     == 'empty'
+-- buildg (\\_ v _ _ -> v x)                                   == 'vertex' x
+-- buildg (\\e v o c -> o ('foldg' e v o c x) ('foldg' e v o c y)) == 'overlay' x y
+-- buildg (\\e v o c -> c ('foldg' e v o c x) ('foldg' e v o c y)) == 'connect' x y
+-- buildg (\\e v o _ -> 'foldr' o e ('map' v xs))                  == 'vertices' xs
+-- buildg (\\e v o c -> 'foldg' e v o ('flip' c) g)                 == 'transpose' g
+-- 'foldg' e v o c (buildg f)                                   == f e v o c
+-- @
+buildg :: (forall b. b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> b) -> Graph a
+buildg f = f Empty Vertex Overlay Connect
+{-# INLINE [1] buildg #-}
 
 composeR :: (b -> c) -> (a -> b) -> a -> c
 composeR = (.)
-{-# INLINE [0] composeR #-}
+{-# INLINE [1] composeR #-}
 
 matchR :: b -> (a -> b) -> (a -> Bool) -> a -> b
 matchR e v p = \x -> if p x then v x else e
```

Comment:
`matchR` doesn't seem to be used in the LHS of a rewrite rule either.

Reply:
You can find it in the `"graph/induce"` rule

Reply:
Indeed, thanks!